### PR TITLE
waitFor(QCoro::Task<T>): Do not require T to be default constructible

### LINKED ISSUE #222: waitFor(QCoro::Task<T>): Do not require T to be default constructible
If you use `QCoro::waitFor` with a `QCoro::Task<T>`, the type `T` must be default constructible. It would nice to be able to use this function with types that do not have a default constructor.

`T` is required to be default constructible because in `impl/waitfor.h:76` a default `T` is constructed, whose value is immediately overwritten in the call to `QCoro::details::runCoroutine()`.

Example compile error from gcc with `T` being a boost::outcome::result<int, std::string>`:

```
In file included from ../../../deps/include/qcoro5/qcoro/qcorotask.h:568,
                 from ../../../deps/include/qcoro5/QCoro/QCoroTask:1,
                 from ../../../test.h:15,
../../../deps/include/qcoro5/qcoro/impl/waitfor.h: In instantiation of 'T QCoro::detail::waitFor(Awaitable&&) [with T = boost::outcome_v2::basic_result<int>, std::__cxx11::basic_string<char>, boost::outcome_v2::policy::exception_ptr_rethrow<int>, std::__cxx11::basic_string<char>, void> >; Awaitable = QCoro::Task<boost::outcome_v2::basic_result<int>, std::__cxx11::basic_string<char>, boost::outcome_v2::policy::exception_ptr_rethrow<int>, std::__cxx11::basic_string<char>, void> > >]':
../../../deps/include/qcoro5/qcoro/impl/waitfor.h:97:30:   required from 'T QCoro::waitFor(QCoro::Task<T>&&) [with T = boost::outcome_v2::basic_result<int>, std::__cxx11::basic_string<char>, boost::outcome_v2::policy::exception_ptr_rethrow<int>, std::__cxx11::basic_string<char>, void> >]'
RemoteCollectorDatabase.h:27:30:   required from here
../../../deps/include/qcoro5/qcoro/impl/waitfor.h:76:11: error: use of deleted function 'boost::outcome_v2::basic_result<R, S, NoValuePolicy>::basic_result() [with R = int>; S = std::__cxx11::basic_string<char>; NoValuePolicy = boost::outcome_v2::policy::exception_ptr_rethrow<int>, std::__cxx11::basic_string<char>, void>]'
   76 |         T result;
      |           ^~~~~~
In file included from ../../../deps/include/boost/outcome/std_result.hpp:34,
                 from ../../../deps/include/boost/outcome/boost_result.hpp:60,
                 from ../../../deps/include/boost/outcome/result.hpp:34,
                 from ../../../libsceptre/test.h:14,
../../../deps/include/boost/outcome/basic_result.hpp:372:3: note: declared here
  372 |   basic_result() = delete;
```

### PR DESCRIPTION
**Pull request by joesdiner**
A patch so that `T` no longer needs to be default constructible when calling `QCoro::waitFor(QCoro::Task<T>)`. Not claiming that this is the best way to do things, but it worked for me. That said I would appreciate any additional testing. I'm only a light user of the QCoro library.

I ended up duplicating some code logic, because I'm not sure if using `std::optional` in all cases could decrease performance. Previously a `T` would always be default constructed, but then RVO should occur when returning the value from `detail::waitFor()` so no additional copy/move would occur. If the default constructor for `T` is cheap, this should have been fine.

With using a `std::optional<T>` though, `T`s default constructor is never called, but I don't think RVO can occur anymore with returning its value from `detail::waitFor()`. I _think_ `T` may be moved to return its value, but this is a corner of C++ I don't have a complete grasp on. 

closes #222

### CODE REVIEW

**Review by danvratil — CHANGES_REQUESTED** [review]
Thanks a lot for the patch! Could you please update the patch and drop the branch for default-constructible types and just always use `std::optional`? It actually is the more correct approach since we avoid the (potentially high) cost of calling the default constructor, which also avoids the triggering of any potential side-effects that the constructor might have.

Could you also please add a unit-test for a non-default-constructible type to make sure we don't regress in the future? Thank you

File: qcoro/impl/waitfor.h [hunk 1]
```diff
@@ -73,15 +86,27 @@ T waitFor(Awaitable &&awaitable) {
...
+        } else {
+            std::optional<T> result;
+            runCoroutineOpt(context, result, std::forward<Awaitable>(awaitable));
+            if (!context.coroutineFinished) {
+                context.loop.exec();
+            }
+            if (context.exception) {
+                std::rethrow_exception(context.exception);
+            }
+            return *result;
```
Comment (danvratil) [inline qcoro/impl/waitfor.h:108]: You should be able to `std::move` the value here. You are right that we lose RVO with the optional, but that should be a negligible cost, especially if we consider the overall overhead of waitFor due to running a nested QEventLoop and suspending/resuming coroutines themselves.

Comment (joesdiner) [reply qcoro/impl/waitfor.h]: added explicit move to hopefully be more effienct

**Comment by joesdiner** [comment]
Uses the `std::optional` paradigm now even `T` has a default constructor. Added some basic unit tests.

**Review by danvratil — APPROVED** [review]
The MacOS CI is lazy today for some reason, but the change passes on Linux Clang and I don't have any reason to suspect AppleClang to not pass as well here.

Thanks a lot for the change and for adding the unittest 👍.

There's one more PR waiting, once it's merged I'll do a new QCoro release with those changes.

**Pull request merged**